traffic_router/core/src/main/java/org/apache/traffic_control/traffic_router/core/loc/AbstractServiceUpdater.java

```diff
@@ -366,13 +369,15 @@ protected File downloadDatabase(final String url, final File existingDb) throws
 		try (InputStream in = conn.getInputStream();
 			 OutputStream out = new FileOutputStream(outputFile)
 		) {
-			eTag = conn.getHeaderField("ETag");
-
-			if (conn.getResponseCode() == HttpURLConnection.HTTP_NOT_MODIFIED) {
-				LOGGER.info("[" + getClass().getSimpleName() + "] " + url + " not modified since our existing database's last update time of " + new Date(existingDb.lastModified()));
-				return existingDb;
+			if (conn instanceof HttpURLConnection) {
+				eTag = conn.getHeaderField("ETag");
+				if (((HttpURLConnection) conn).getResponseCode() == HttpURLConnection.HTTP_NOT_MODIFIED) {
```

Comment:
If `conn` is a `FileURLConnection`, it should check `conn.getLastModified()` against `existingDb.lastModified()` and return the existing DB if they are the same.

Reply:
Where's the definition of `FileURLConnection` - I've seen it in stack traces as `sun.net.www.protocol.file.FileURLConnection`, but trying to import that causes compilation to fail with:

```
error: package sun.net.www.protocol.file is not visible
```

Could that possibly be an implementation detail of the JRE used?

Reply:
Yeah, that class is not visible. Technically, it could be made visible by adding
```xml
					<compilerArgs>
						<arg>-XDignore.symbol.file</arg>
						<arg>--add-exports</arg>
						<arg>java.base/sun.net.www.protocol.file=ALL-UNNAMED</arg>
					</compilerArgs>
```
to the `maven-compiler-plugin` configuration options in place of https://github.com/apache/trafficcontrol/blob/739124d0318b184185ecf2f80cbbc12c6c36e0ba/traffic_router/core/pom.xml#L148

but doing that results in a PMD failure:
```java
[INFO] PMD Failure: org.apache.traffic_control.traffic_router.core.loc.AbstractServiceUpdater:36 Rule:DontImportSun Priority:4 Avoid importing anything from the sun.* packages.
```
Instead of checking for `instanceof FileURLConnection`, "not `instanceof HttpURLConnection`" should be good enough, IMO

Reply:
I was gonna say I'll check the protocol string in comparison to `file` (or `file:` or whatever it is for `FileURLConnection`s), but since `getLastModified` is a method of `URLConnection` I can't imagine what difference it would make.

On the other hand, the docs for `URLConnection.getLastModified` say
> Returns the value of the `last-modified` header field.

(Although I think I was looking at Java 8 for that). Does the `FileURIConnection` sub class implement that with modification timestamps on supported filesystems. What about on filesystems that don't store that information?

Reply:
> I was gonna say I'll check the protocol string in comparison to `file` (or `file:` or whatever it is for `FileURLConnection`s)

That sounds right, checking the protocol is how the URL class decides which handler to use.

> (Although I think I was looking at Java 8 for that). Does the `FileURIConnection` sub class implement that with modification timestamps on supported filesystems. What about on filesystems that don't store that information?

Maybe it returns 0 in those cases? When we don't have that info, we can skip that check. We use `useModifiedTimestamp()` to check this case for HTTP URLs. We can use it for File URLs too, right?

https://github.com/apache/trafficcontrol/blob/de41a02eb357de3caebb692804213776884f1174/traffic_router/core/src/main/java/org/apache/traffic_control/traffic_router/core/loc/AbstractServiceUpdater.java#L387-L390

Reply:
> We can use it for File URLs too, right?

¯\\\_(ツ)\_/¯ as far as I know, but that seems to be up to the implementation of that URL protocol subclass.